modules/samples/virtual_firealarm/src/org.wso2.carbon.device.mgt.iot.sample.virtual.firealarm.service.impl/src/main/java/org/wso2/carbon/device/mgt/iot/sample/virtual/firealarm/service/impl/util/xmpp/XMPPClient.java

```diff
@@ -0,0 +1,241 @@
+	 * @param password the password of the device's XMPP-Account.
+	 * @param resource the resource the resource, specific to the XMPP-Account to which the login
+	 *                 is made to
+	 * @throws DeviceManagementException in the event of 'Connecting to' or 'Logging into' the
+	 *                                     XMPP server fails.
+	 */
+	public void connectAndLogin(String username, String password, String resource)
+			throws DeviceManagementException {
+		try {
+			connection.connect();
+			log.info(String.format(
+					"Connection to XMPP Server at %1$s established successfully......", server));
+
+		} catch (XMPPException xmppExcepion) {
+			String errorMsg =
+					"Connection attempt to the XMPP Server at " + server + " via port " + port +
+							" failed.";
+			log.info(errorMsg);
+			throw new DeviceManagementException(errorMsg, xmppExcepion);
+		}
+
+		if (connection.isConnected()) {
+			try {
+				if (resource == null) {
+					connection.login(username, password);
+					log.info(String.format("Logged into XMPP Server at %1$s as user %2$s......", server, username));
+				} else {
+					connection.login(username, password, resource);
+					log.info(String.format(
+							"Logged into XMPP Server at %1$s as user %2$s on resource %3$s......",
+							server, username, resource));
+				}
+			} catch (XMPPException xmppException) {
+				String errorMsg =
+						"Login attempt to the XMPP Server at " + server + " with username - " +
+								username + " failed.";
+				log.info(errorMsg);
+				throw new DeviceManagementException(errorMsg, xmppException);
+			}
+		}
+	}
+
+	/**
+	 * Sets a filter on all the incoming XMPP-Messages for the JID (XMPP-Account ID) passed in.
+	 * Also creates a listener for the incoming messages and connects the listener to the
+	 * XMPPConnection alongside the set filter.
+	 *
+	 * @param senderJID the JID (XMPP-Account ID) to which the filter is to be set.
+	 */
+	public void setMessageFilterAndListener(String senderJID) {
+		filter = new AndFilter(new PacketTypeFilter(Message.class), new
+				FromContainsFilter(
+				senderJID));
+		listener = new PacketListener() {
+			@Override
+			public void processPacket(Packet packet) {
+				if (packet instanceof Message) {
+					final Message xmppMessage = (Message) packet;
+					Thread msgProcessThread = new Thread() {
+						public void run() {
+							processXMPPMessage(xmppMessage);
+						}
+					};
+					msgProcessThread.start();
+				}
+			}
+		};
+
+		connection.addPacketListener(listener, filter);
+	}
+
+
+
+	/**
+	 * Sends an XMPP message
+	 *
+	 * @param JID     the JID (XMPP Account ID) to which the message is to be sent to.
+	 * @param message the XMPP-Message that is to be sent.
+	 */
+	public void sendXMPPMessage(String JID, String message) {
+		sendXMPPMessage(JID, message, "Reply-From-Device");
+		if (log.isDebugEnabled()) {
+			log.debug("Message: " + message + " to XMPP JID [" + JID + "] sent successfully");
+		}
+	}
+
```

Comment:
what is the rational behind having two similar methods here to do the same? only difference is having or not having a subject...